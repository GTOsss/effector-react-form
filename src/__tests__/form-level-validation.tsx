import { Controller, OnSubmit } from '../../index';
import React from 'react';
import { useForm } from '../index';
import { render, screen, fireEvent } from '@testing-library/react';

interface Values {
  username?: string;
  profile?: {
    firstName?: string;
    lastName?: string;
  };
}

interface InputProps {
  controller: Controller;
  label: string;
}

const formValidate = ({ values }) => {
  const errors: Record<string, any> = {};

  if (!values.username) {
    errors.username = 'Field is required';
  } else if (values.username.length < 4) {
    errors.username = 'Minimum of 4 characters';
  }

  if (!values.profile || !values.profile.firstName) {
    errors['profile.firstName'] = 'Field is required'; // Field without nesting!
  }

  return errors;
};

const Input: React.FC<InputProps> = ({ controller, label }) => {
  const { input, fieldState, form, error } = controller();

  const isShowError = error && (form.submitted || form.hasOuterError || fieldState.blurred);

  return (
    <div role="wrapper-for-input" className="input-wrap">
      <label>{label}</label>
      <input
        {...input}
        value={input.value || ''}
        role="textbox"
        className={`input${isShowError ? ' input-error' : ''}`}
        placeholder={label}
      />
      {isShowError && <span>{error}</span>}
    </div>
  );
};

const FieldLevelValidation = () => {
  const { handleSubmit, controller, setOrDeleteError } = useForm({ validate: formValidate, onSubmit: () => {} });

  return (
    <form onSubmit={handleSubmit}>
      <Input label="Username" controller={controller({ name: 'username' })} />
      <Input label="First name" controller={controller({ name: 'profile.firstName' })} />
      <Input label="Last name" controller={controller({ name: 'profile.lastName' })} />
      <button type="submit">submit</button>
      <button
        type="button"
        onClick={() =>
          setOrDeleteError({
            field: 'profile.firstName',
            error: 'First name is not valid',
          })
        }
      >
        set error for firstName
      </button>
    </form>
  );
};

describe('FormLevelValidation', () => {
  test('submit', () => {
    render(<FieldLevelValidation />);
    const buttonSubmit = screen.getByText('submit');
    fireEvent.click(buttonSubmit);
    const inputs = screen.getAllByRole('wrapper-for-input');
    expect(inputs).toMatchSnapshot();
  });

  test('onFocus, onBlur', () => {
    render(<FieldLevelValidation />);
    const inputFirstName = screen.getByPlaceholderText('First name');
    fireEvent.focus(inputFirstName);
    fireEvent.blur(inputFirstName);
    const inputs = screen.getAllByRole('wrapper-for-input');
    expect(inputs).toMatchSnapshot();
  });

  test('onFocus, onChange: "t" -> ""', () => {
    render(<FieldLevelValidation />);
    const inputFirstName = screen.getByPlaceholderText('First name');
    fireEvent.focus(inputFirstName);
    fireEvent.change(inputFirstName, { target: { value: 't' } });
    fireEvent.change(inputFirstName, { target: { value: '' } });
    const inputs = screen.getAllByRole('wrapper-for-input');
    expect(inputs).toMatchSnapshot();
  });

  test('onFocus, onChange: "t" -> "", onBlur', () => {
    render(<FieldLevelValidation />);
    const inputFirstName = screen.getByPlaceholderText('First name');
    fireEvent.focus(inputFirstName);
    fireEvent.change(inputFirstName, { target: { value: 't' } });
    fireEvent.change(inputFirstName, { target: { value: '' } });
    fireEvent.blur(inputFirstName);
    const inputs = screen.getAllByRole('wrapper-for-input');
    expect(inputs).toMatchSnapshot();
  });
});
