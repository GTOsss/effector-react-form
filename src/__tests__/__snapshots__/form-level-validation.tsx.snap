// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLevelValidation onFocus, onBlur 1`] = `
Array [
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Username
    </label>
    <input
      class="input"
      name="username"
      placeholder="Username"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      First name
    </label>
    <input
      class="input"
      name="profile.firstName"
      placeholder="First name"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Last name
    </label>
    <input
      class="input"
      name="profile.lastName"
      placeholder="Last name"
      role="textbox"
      value=""
    />
  </div>,
]
`;

exports[`FormLevelValidation onFocus, onChange: "t" -> "" 1`] = `
Array [
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Username
    </label>
    <input
      class="input"
      name="username"
      placeholder="Username"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      First name
    </label>
    <input
      class="input"
      name="profile.firstName"
      placeholder="First name"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Last name
    </label>
    <input
      class="input"
      name="profile.lastName"
      placeholder="Last name"
      role="textbox"
      value=""
    />
  </div>,
]
`;

exports[`FormLevelValidation onFocus, onChange: "t" -> "", onBlur 1`] = `
Array [
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Username
    </label>
    <input
      class="input"
      name="username"
      placeholder="Username"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      First name
    </label>
    <input
      class="input"
      name="profile.firstName"
      placeholder="First name"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Last name
    </label>
    <input
      class="input"
      name="profile.lastName"
      placeholder="Last name"
      role="textbox"
      value=""
    />
  </div>,
]
`;

exports[`FormLevelValidation submit 1`] = `
Array [
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Username
    </label>
    <input
      class="input"
      name="username"
      placeholder="Username"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      First name
    </label>
    <input
      class="input"
      name="profile.firstName"
      placeholder="First name"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Last name
    </label>
    <input
      class="input"
      name="profile.lastName"
      placeholder="Last name"
      role="textbox"
      value=""
    />
  </div>,
]
`;
